addons/l10n_cl/models/account_move.py

```diff
@@ -27,23 +26,92 @@ def _get_l10n_latam_documents_domain(self):
-                if self.partner_id.l10n_cl_sii_taxpayer_type == '2':
-                    domain += [('code', '=', '71')]
-                return domain
-            document_type_ids = self.journal_id.l10n_cl_sequence_ids.mapped('l10n_latam_document_type_id').ids
+            if self.journal_id.type == 'sale':
+                document_type_ids = self.journal_id.l10n_cl_sequence_ids.mapped('l10n_latam_document_type_id').ids
+            else:  # self.journal_id.type == 'purchase':
+                partner_domain = [
+                        ('country_id.code', '=', 'CL'),
+                        ('internal_type', 'in', ['invoice', 'debit_note', 'credit_note']), ('active', '=', True)]
+                if len(self.partner_id) == 0:
+                    # this will allow to use the whole documents domain if partner not selected.
+                    pass
+                elif self.partner_id.l10n_cl_sii_taxpayer_type == '1':
+                    partner_domain += [('code', 'not in', ['39', '70', '71'])]
+                elif self.partner_id.l10n_cl_sii_taxpayer_type == '2':
+                    partner_domain += [('code', 'in', ['70', '71', '56', '61'])]
+                elif self.partner_id_vat == '60805000-0':
+                    partner_domain += [('code', '=', '914')]
+                elif self.partner_id.l10n_cl_sii_taxpayer_type == '3':
+                    partner_domain += [('code', 'in', ['35', '38', '39', '41', '56', '61'])]
+                elif not self.partner_id.l10n_cl_sii_taxpayer_type or self.partner_id.country_id != self.env.ref(
+                        'base.cl') or self.partner_id.l10n_cl_sii_taxpayer_type == '4':
+                    partner_domain += [('code', 'in', [])]
+                document_type_obj = self.env['l10n_latam.document.type'].search(partner_domain)
+                document_type_ids = document_type_obj.ids
             domain += [('id', 'in', document_type_ids)]
-            if self.partner_id.l10n_cl_sii_taxpayer_type == '3':
-                domain += [('code', 'in', ['35', '38', '39', '41', '56', '61'])]
         return domain
 
-    @api.constrains('type', 'l10n_latam_document_type_id')
+    @api.depends('journal_id', 'partner_id', 'company_id', 'type')
+    def _compute_l10n_latam_available_document_types(self):
+        for record in self:
+            if record.journal_id.l10n_latam_use_documents and \
+                    record.journal_id.company_id.country_id == self.env.ref('base.cl'):
+                record.l10n_latam_available_document_type_ids = False
+                for rec in self.filtered(lambda x: x.journal_id and x.l10n_latam_use_documents):
+                    rec.l10n_latam_available_document_type_ids = self.env['l10n_latam.document.type'].search(
+                        rec._get_l10n_latam_documents_domain())
+            else:
+                super()._compute_l10n_latam_available_document_types()
+
+    @api.constrains('state', 'type', 'l10n_latam_document_type_id')
```

Comment:
I would prefer these checks on the post.  A reason is e.g. that it depends on the partner (e.g. tax_payer_type), but it should not be triggered by it. The only moment we really want to do this validation is at the post.  (e.g. not afterwards)  It is also one trigger less for the orm to take into account.

Reply:
ok

Reply:
We execute the method inside the post and remove api constrains at

ccad6d8f3bb8732756a53adc260245de9eee06b4